### PR TITLE
use git ignore rules to skip hooks

### PR DESCRIPTION
**Pull request by nik-kor**

### CODE REVIEW

**Comment by tarmolov** [comment]
Now you probably start hating me ^^

I think `.githooksignore` should absolutely the same as original [.gitignore](https://git-scm.com/docs/gitignore) because `git-hooks` and `git` are tightly connected and tangled ;)

**Comment by Guria** [comment]
@tarmolov +1 
@nik-kor the best place to store such ignaores (os, editor specific) is your global `.gitttgnore`

**Comment by nik-kor** [comment]
@tarmolov I hate you. No, just joking)

yeah, it's a good idea to use global ignore settings. Have to figure out how to extract them because they can be anywhere and I, probably, should to analyze git settings

**Comment by Guria** [comment]
@nik-kor why ever need to extract it. Just manage it on your own machine.

**Comment by nik-kor** [comment]
@Guria because I want to teach git-hooks to ignore some files inside git-hooks/ folders.

**Comment by Guria** [comment]
Aren't githooks already ignores all that ignored by git just from its nature?

**Comment by nik-kor** [comment]
No, it isn't.) 
I mean this https://github.com/tarmolov/git-hooks-js/blob/master/lib/git-hooks.js#L109 should be teached to ignore some stuff from .githooks

**Comment by tarmolov** [comment]
Let's use [git-check-ignore](https://git-scm.com/docs/git-check-ignore) to check path before execution hook. It seems it'll solve the issue and you don't need to add a separate `.githooksignore`.

```shell
$ git check-ignore -q node_modules/.bin/mocha
$ echo $?
0
$ git check-ignore -q bin/git-hooks
$ echo $?
1
```

**Comment by nik-kor** [comment]
have to check while it is not working in node 0.10....(

File: lib/git-hooks.js [hunk 1]
```diff
@@ -3,6 +3,12 @@ var util = require('util');
 var spawn = require('child_process').spawn;
 var fs = require('fs');
 var fsHelpers = require('./fs-helpers');
+var execSync = require('child_process').execSync;
+if (!execSync) {
+    execSync = require('runsync').exec;
```
Comment (tarmolov) [inline lib/git-hooks.js]: External modules are forbidden.

Comment (nik-kor) [inline lib/git-hooks.js]: sorry, can you explain why?

Comment (tarmolov) [inline lib/git-hooks.js]: because of [npm@2 bug](https://github.com/npm/npm/issues/8850) :-/

File: lib/git-hooks.js [hunk 2]
```diff
@@ -111,6 +117,22 @@ module.exports = {
                 return path.resolve(hooksDirname, hookName);
             });
 
+            var ignoreList = [];
```
Comment (tarmolov) [inline lib/git-hooks.js]: First of all, please move this code to the separate function. Let's call it `exludeIgnoredPathes`, for example.
```
/**
  * @param {String[]} hooks List of hook names to execute.
  * @returns {String[]} hooksWithoutIgnoredPathes
  **/ 
function exludeIgnoredPathes(hooks) {
}
```

File: lib/git-hooks.js [hunk 3]
```diff
@@ -111,6 +117,22 @@ module.exports = {
...
+            try {
```
Comment (tarmolov) [inline lib/git-hooks.js]: When does it throw an error?

File: lib/git-hooks.js [hunk 4]
```diff
@@ -111,6 +117,22 @@ module.exports = {
...
+                ignoreList = execSync(
+                    'git check-ignore ' +
+                    hooks
+                        .map(function (hook) { return path.resolve(hooksDirname, hook); })
```
Comment (tarmolov) [inline lib/git-hooks.js]: It's already done in line 117.

File: lib/git-hooks.js [hunk 5]
```diff
@@ -111,6 +117,22 @@ module.exports = {
...
+                        .join(' ')
+                ).toString().split(EOL).filter(function (path) { return path; });
```
Comment (tarmolov) [inline lib/git-hooks.js]: You don't need to split the result. `indexOf` works with strings as well.

File: lib/git-hooks.js (hunk 1 above)
Comment (nik-kor) [inline lib/git-hooks.js]: maybe I'm dumb, but I haven't catched it yet.)
So the problem you mentioned is with postinstall npm-script? 
And here we just reassign variable so polyfilling it.

You mean that this
```
"preinstall": "if [ $(node -e 'console.log(/0.10/.test(process.version))') = 'true' ]; then npm install runsync; else exit 0; fi"
```
script is buggy?

So we can just always install runsync module

File: lib/git-hooks.js (hunk 3 above)
Comment (nik-kor) [inline lib/git-hooks.js]: when no ignores found

File: lib/git-hooks.js (hunk 4 above)
Comment (nik-kor) [inline lib/git-hooks.js]: yep)

File: lib/git-hooks.js (hunk 5 above)
Comment (nik-kor) [inline lib/git-hooks.js]: yeah. definetly)

File: lib/git-hooks.js (hunk 1 above)
Comment (tarmolov) [inline lib/git-hooks.js]: I really don't like this `preinstall` script. Let's implement it without an external module. Async version is fine.

File: lib/git-hooks.js (hunk 3 above)
Comment (tarmolov) [inline lib/git-hooks.js]: Just check `.gitignore` existence before running command and you don't need `try..catch` at all.

Comment (nik-kor) [inline lib/git-hooks.js]: I believe that git-check-ignore use .gitignore from the home directory. So if we could just read .gitignore from project directory there won't be the case for git-check-ignore command

File: lib/git-hooks.js (hunk 1 above)
Comment (nik-kor) [inline lib/git-hooks.js]: What if we install runsync always?


I just use sync version because this test fails
```
  1) git-hook runner when hooks are found and a hook is unexecutable should return an error:
     AssertionError: expected [Function] to throw Error
```
if you think that async exec will be better than I'll rewrite the code.

Comment (tarmolov) [inline lib/git-hooks.js]: I do love sync version but I want to support `0.8` and `0.10` as well without extra modules. Let's make async version.
```javascript
excludeIgnoredPaths(hooks, function (filteredHooks) {
    runHooks(filteredHooks, [arg], callback);
})
```

Comment (nik-kor) [inline lib/git-hooks.js]: sure.

File: lib/hook-template.js [hunk 6]
```diff
@@ -1,9 +1,8 @@
 #!/usr/bin/env node
 
-try {
```
Comment (tarmolov) [inline lib/hook-template.js]: 1. Install hooks.
2. Remove `node_modules` directory.
-> you cannot make a commit ;)

`try..catch` saves the day here.

File: tests/run.test.js [hunk 7]
```diff
@@ -42,11 +44,12 @@ describe('git-hook runner', function () {
...
 
-            it('should return an error', function () {
-                var fn = function () {
-                    gitHooks.run(PRECOMMIT_HOOK_PATH);
-                };
-                fn.should.throw(Error);
+            it('should return an error', function (done) {
+                gitHooks.run(PRECOMMIT_HOOK_PATH, null, function (code, error) {
+                    code.should.equal(1);
+                    expect(error).to.be.ok;
```
Comment (tarmolov) [inline tests/run.test.js]: error.should.be.ok

File: lib/git-hooks.js [hunk 8]
```diff
@@ -131,14 +146,18 @@ function runHooks(hooks, args, callback) {
...
 
-    var hook = spawnHook(hooks.shift(), args);
-    hook.on('close', function (code) {
-        if (code === 0) {
-            runHooks(hooks, args, callback);
-        } else {
-            callback(code);
-        }
-    });
+    try {
```
Comment (tarmolov) [inline lib/git-hooks.js]: Let's revert this changes.

File: lib/git-hooks.js [hunk 9]
```diff
@@ -25,6 +26,19 @@ var HOOKS = [
     'update'
 ];
 
+/**
+ * @param {String[]} list of hook's paths with possible excludes(.gitignore files)
+ * @param  {Function} callback - passing filtered hooks
```
Comment (tarmolov) [inline lib/git-hooks.js]: Please fix jsdoc here
```javascript
* @param {String[]} hooks List of hook's paths with possible excludes (.gitignore files)
* @param  {Function} callback Filtered hooks will be passed in the callback
```

**Comment by tarmolov** [comment]
Just a couple of remarks and I'm ready to merge it :)

File: lib/git-hooks.js (hunk 8 above)
Comment (nik-kor) [inline lib/git-hooks.js]: try-catch here is required because we cannot catch the error in the hook(from hook-template) because the code became asynchronous -  excludeIgnoredPaths is async now.
That's why I tried to use sync version of exec

File: lib/hook-template.js (hunk 6 above)
Comment (nik-kor) [inline lib/hook-template.js]: try-catch is possible here if we use sync version of exec.

> Remove node_modules directory. -> you cannot make a commit ;)
what do you mean?)

Comment (tarmolov) [inline lib/hook-template.js]:
```sh
$ mkdir test
$ cd test
$ git init
Initialized empty Git repository in /Users/hevil/test/.git/
$ npm install git-hooks
$ touch README.md
$ git commit -am "Initial commit"
$ cat .git/hooks/pre-commit
#!/usr/bin/env node

require('../../node_modules/git-hooks/lib/git-hooks').run(__filename, process.argv[2], function (code) {
    process.exit(code);
});
$ rm -r node_modules
$ echo "test" > README.md
$ git commit -am "Commit which fails"
module.js:338
    throw err;
          ^
Error: Cannot find module '../../node_modules/git-hooks/lib/git-hooks'
    at Function.Module._resolveFilename (module.js:336:15)
    at Function.Module._load (module.js:278:25)
    at Module.require (module.js:365:17)
    at require (module.js:384:17)
    at Object.<anonymous> (/Users/hevil/test/.git/hooks/pre-commit:3:1)
    at Module._compile (module.js:460:26)
    at Object.Module._extensions..js (module.js:478:10)
    at Module.load (module.js:355:32)
    at Function.Module._load (module.js:310:12)
    at Function.Module.runMain (module.js:501:10)
```

File: lib/git-hooks.js (hunk 8 above)
Comment (tarmolov) [inline lib/git-hooks.js]: I don't see `execSync` in your patch.

Comment (nik-kor) [inline lib/git-hooks.js]: i removed it because https://github.com/tarmolov/git-hooks-js/pull/14#discussion_r53565778

Comment (nik-kor) [inline lib/git-hooks.js]: so we experimented with async version and it sucks. Let use the sync one ?)

Comment (tarmolov) [inline lib/git-hooks.js]: It doesn't suck actually.

Sync version is bad because of two reasons:
  * if a hook works a long time, a user will see the output when the hook ends.
  * You have to specify an enormous size of buffer `maxBuffer` because you don't know how many output the hook prints.

File: lib/hook-template.js (hunk 6 above)
Comment (nik-kor) [inline lib/hook-template.js]: I'm dumb....) 

fixed it

**Comment by nik-kor** [comment]
rebased it.

**Comment by tarmolov** [comment]
Thank you

**Comment by nik-kor** [comment]
Thank you too! 
I knew some stuff doing this tiny PR)
